src/ring_theory/polynomial/chebyshev/dickson.lean

```diff
@@ -0,0 +1,127 @@
+
+/-!
+# Dickson polynomials
+
+The Dickson polynomials are two families of polynomials indexed by `ℕ`,
+with coefficients in a commutative ring `R` depending on an element `a∈R`.
+They are closely related to the Chebyshev polynomials in the case that `a=1`.
+When `a=0` they are just the family of monomials `x ^ n`.
+
+## Main declarations
+
+* `polynomial.dickson₁`: the Dickson polynomials of the first kind.
+* `polynomial.dickson₂`: the Dickson polynomials of the second kind.
+
+## References
+
+* [R. Lidl, G. L. Mullen and G. Turnwald, _Dickson polynomials_][MR1237403]
+
+## TODO
+
+* Move the definition of `lambdashev` from `chebyshev.defs` into this file
+  (or even remove it because it is a special case of the polynomials defined here?).
+* Show that `lambdashev` is equal to the characteristic polynomial of the adjacency matrix of a
+  type A Dynkin diagram.
+-/
+
+noncomputable theory
+
+namespace polynomial
+
+variables (R S : Type*) [comm_ring R] [comm_ring S] (a : R)
+
+/-- `dickson₁` is the `n`-th Dickson polynomial of the first kind associated to the element `a∈R`.
+-/
+noncomputable def dickson₁ : ℕ → polynomial R
+| 0       := 2
+| 1       := X
+| (n + 2) := X * dickson₁ (n + 1) - (monomial 0 a) * dickson₁ n
+
+@[simp] lemma dickson₁_zero : dickson₁ R a 0 = 2 := rfl
+@[simp] lemma dickson₁_one : dickson₁ R a 1 = X := rfl
+lemma dickson₁_two : dickson₁ R a 2 = X ^ 2 - monomial 0 a * 2 :=
+by simp only [dickson₁, mul_one, pow_two]
+@[simp] lemma dickson₁_add_two (n : ℕ) :
+  dickson₁ R a (n + 2) = X * dickson₁ R a (n + 1) - (monomial 0 a) * dickson₁ R a n :=
+by rw dickson₁
+
+lemma dickson₁_of_two_le (n : ℕ) (h : 2 ≤ n) :
+  dickson₁ R a n = X * dickson₁ R a (n - 1) - (monomial 0 a) * dickson₁ R a (n - 2) :=
+begin
+  obtain ⟨n, rfl⟩ := nat.exists_eq_add_of_le h,
+  rw add_comm,
+  exact dickson₁_add_two R a n
+end
+
+variables {R S a}
+
+lemma map_dickson₁ (f : R →+* S) :
+  ∀ (n : ℕ), map f (dickson₁ R a n) = dickson₁ S (f a) n
+| 0       := by simp only [dickson₁_zero, bit0, map_add, map_one]
+| 1       := by simp only [dickson₁_one, map_X]
+| (n + 2) :=
+begin
+  simp only [dickson₁_add_two, map_sub, map_mul, map_X, map_monomial],
+  rw [map_dickson₁ (n + 1), map_dickson₁ n]
+end
+
+variable (R)
+
+lemma lambdashev_eq_dickson₁_of_eq_one :
+∀ (n : ℕ), lambdashev R n = dickson₁ R 1 n
+| 0       := by simp only [lambdashev_zero, dickson₁_zero]
+| 1       := by simp only [lambdashev_one, dickson₁_one]
+| (n + 2) :=
+begin
+  simp only [lambdashev_add_two, dickson₁_add_two],
+  rw [lambdashev_eq_dickson₁_of_eq_one, lambdashev_eq_dickson₁_of_eq_one],
+  change X * dickson₁ R 1 (n + 1) - dickson₁ R 1 n = X * dickson₁ R 1 (n + 1) - 1 * dickson₁ R 1 n,
+  rw one_mul
+end
+
+variables (R S a)
+
+/-- `dickson₂` is the `n`-th Dickson polynomial of the second kind associated to the element `a∈R`.
+-/
+noncomputable def dickson₂ : ℕ → polynomial R
+| 0       := 1
+| 1       := X
+| (n + 2) := X * dickson₂ (n + 1) - (monomial 0 a) * dickson₂ n
```

Comment:
This looks exactly the same as `dickson₁`, but with `2` replaced by 1. If so, can you generalize to
```lean
noncomputable def dickson' (a₀ : R) : ℕ → polynomial R
| 0       := a₀
| 1       := X
| (n + 2) := X * dickson' (n + 1) - (monomial 0 a) * dickson' n

abbrevation dickson₁ := dickson' a 2
abbrevation dickson₂ := dickson' a 1
```
which probably halves the number of proofs you have to write.

Reply:
I agree that it could save a number of proofs but even after some searching I have not found any mention of this generalisation of the Dickson polynomials.

Reply:
Deviating from published sources to make formalization easier is perfectly valid, provided you're still able to express everything needed later / provide a "compatible API". It might be worth pointing out nonstandard definitions in the doc strings and the module doc string of course.

Reply:
> This looks exactly the same as `dickson₁`, but with `2` replaced by 1. If so, can you generalize to
> 
> ```lean
> noncomputable def dickson' (a₀ : R) : ℕ → polynomial R
> | 0       := a₀
> | 1       := X
> | (n + 2) := X * dickson' (n + 1) - (monomial 0 a) * dickson' n
> 
> abbreviation dickson₁ := dickson' a 2
> abbreviation dickson₂ := dickson' a 1
> ```
> 
> which probably halves the number of proofs you have to write.

This suggestion seems to not work so nicely as there seems to be no coercion from R to polynomial R. Instead I would have to use monomial 0 a₀ which makes proofs more tedious. I could try to use (a₀ : nat) instead.

Reply:
I tried it with this suggestion but at least for me it was not easier. Feel free to start an attempting maybe proving one of the cases. (I ran into trouble as if I understand correctly the a is now implicit and the property one might want to prove at some point is dickson1 a (n*m) (X) = dickson1 a^n m (dickson1 a n).

Reply:
@eric-wieser suggested to instead define dickson1 and dickson2 in terms of `linear_recurrence`. 

`import algebra.linear_recurrence
import data.matrix.notation

def dickson1_rec (a : R) : linear_recurrence (polynomial R) :=
⟨2, ![C a, X]⟩

def dickson1 (a : R) := (dickson_rec a).mk_sol ![2, X]` 

I would like to do that instead of using an artifical dickson'-function to connect to the general theory. Unfortunately, this is beyond my knowledge of lean at the moment. I added it to the TODO.

Reply:
> I tried it with this suggestion but at least for me it was not easier. Feel free to start an attempting maybe proving one of the cases. (I ran into trouble as if I understand correctly the a is now implicit and the property one might want to prove at some point is dickson1 a (n*m) (X) = dickson1 a^n m (dickson1 a n).

Why not just

```
noncomputable def dickson' (a₀ : polynomial R) : ℕ → polynomial R
| 0       := a₀
| 1       := X
| (n + 2) := X * dickson' (n + 1) - (monomial 0 a) * dickson' n

abbreviation dickson₁ := dickson' a 2
abbreviation dickson₂ := dickson' a 1
```

Then there's no coercion necessary, and in the definition of `dickson₁` and `dickson₂` Lean can see the numeral should be interpreted as a polynomial.

Reply:
Moreover isn't https://en.wikipedia.org/wiki/Dickson_polynomial#Generalization exactly this generalization?

Reply:
I now implemented the wikipedia generalisation, which turns out nicely as one can write dickson 1 and dickson 2 instead of dickson\1 and dickson\2 and state each lemma just once instead of twice. Occasionally, one has to insert an additional norm_num or cast because of that, but it doesn't seem too bad.